src/main/cpp/src/hive_hash.cu

```diff
@@ -486,16 +529,70 @@ std::unique_ptr<cudf::column> hive_hash(cudf::table_view const& input,
 
   check_nested_depth(input);
 
+  // `flattened_column_views` only contains nested columns and columns that result from flattening
+  // nested columns
+  std::vector<cudf::column_view> flattened_column_views;
+  // `first_child_index` has the same size as `flattened_column_views`
+  std::vector<cudf::size_type> first_child_index;
+  // `nested_column_map` has the same size as `input.num_columns()`, and it maps the column index in
+  // `input` to the index in `flattened_column_views`
+  std::vector<cudf::size_type> nested_column_map;
+
+  for (auto const& root_col : input) {
+    if (root_col.type().id() == cudf::type_id::LIST ||
+        root_col.type().id() == cudf::type_id::STRUCT) {
+      // Construct the `flattened_column_views` by level order traversal
+      nested_column_map.push_back(flattened_column_views.size());
+      flattened_column_views.push_back(root_col);
```

Comment:
As mentioned earlier, if we need placeholder for the nested columns, we need to store empty nested columns instead of the original nested ones.
```suggestion
      auto const empty_nested_col = root_col.type().id() == cudf::type_id::LIST ? 
          cudf::column_view{ /* create empty lists column_view */} :
          cudf::column_view{/* create empty structs column_view */};
      flattened_column_views.push_back(empty_nested_col);
```

Reply:
But if I `push_back` `empty_nested_col`, I will not be able to process the children of the `empty_nested_col`.

The element type of `flattened_column_views` is `column_view`, while the element type of `flattened_column_device_views` is `column_device_view`.

Do you mean to store empty nested columns in `flattened_column_device_views`? 👀

Reply:
`flattened_column_views` stores `column_view` which will be transformed into `column_device_view`.

In addition to the current arrays, you also need an array storing the number of children columns for each flattened columns. Using it, you will be able to process the children columns of the flattened nested columns. The empty nested columns are just placeholder so you can detect if a column is nested.

Reply:
If using `num_children` array, we will not need the `first_child_index` array. Since we are flattening by pre-order visits, we always have index of the first child column next to the current column index. We then only need the number of children to know how far to travel from the current index.

Reply:
By the way, @res-life mentioned that creating a `column_view` will call a kernel to calculate the null count, which can be time-consuming. Is there a way to avoid creating `column_views` and directly construct `column_device_views`?

Reply:
> If using `num_children` array, we will not need the `first_child_index` array. Since we are flattening by pre-order visits, we always have index of the first child column next to the current column index. We then only need the number of children to know how far to travel from the current index.

I am using level-order traversal, because it ensures that the children are stored contiguously. 
I think that `first_child_index` can't be removed. For example, with input `STRUCT1<STRUCT2<INT, STRING>, BOOL>`:
```
flattened_columns = [STRUCT1, STRUCT2, BOOL, INT, STRING]
```
The first child(`INT`) of `STRUCT2` is not next to `STRUCT2`.

Reply:
> By the way, @res-life mentioned that creating a `column_view` will call a kernel to calculate the null count, which can be time-consuming. Is there a way to avoid creating `column_views` and directly construct `column_device_views`?

You are creating an array of `column_device_view` from a source array of `column_view` using `contiguous_copy_column_device_views`. Changing  nested `column_view` in that source array won't affect anything.

Reply:
> I am using level-order traversal, because it ensures that the children are stored contiguously.

The children don't have to be stored continuously. They are just stored after their parent. Reconstructing the nested children is similar to this:
```
operator()(...., int& start_index) {
  auto const col = flattened_cviews[start_index];
  auto const num_children = num_childrens[start_index];
  ++start_index;
}

{
  // usage:
  int start_index = 0;
  type_dispatcher(..., element_hasher_adapter{}, start_index);

```